crates/era-downloader/src/stream.rs

```diff
@@ -0,0 +1,216 @@
+use crate::EraClient;
+use futures_util::{stream::FuturesOrdered, FutureExt, Stream, StreamExt};
+use pin_project::pin_project;
+use reqwest::Url;
+use std::{
+    collections::VecDeque,
+    fmt::{Debug, Formatter},
+    future::Future,
+    path::Path,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// An asynchronous stream of ERA1 files.
+#[derive(Debug)]
+pub struct EraStream {
+    download_stream: DownloadStream,
+    starting_stream: StartingStream,
+}
+
+impl EraStream {
+    /// Constructs a new [`EraStream`] that downloads concurrently up to `max_concurrent_downloads`
+    /// ERA1 files to `client` `folder`, keeping their count up to `max_files`.
+    pub fn new(client: EraClient, max_files: usize, max_concurrent_downloads: usize) -> Self {
+        Self {
+            download_stream: DownloadStream {
+                downloads: Default::default(),
+                scheduled: Default::default(),
+                max_concurrent_downloads,
+                ended: false,
+            },
+            starting_stream: StartingStream {
+                client,
+                files_count: Box::pin(async move { usize::MAX }),
+                next_url: Box::pin(async move { Ok(None) }),
+                recover_index: Box::pin(async move { 0 }),
+                state: Default::default(),
+                max_files,
+                max_concurrent_downloads,
+                index: 0,
+                downloading: 0,
+            },
+        }
+    }
+}
+
+impl Stream for EraStream {
+    type Item = eyre::Result<Box<Path>>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if let Poll::Ready(fut) = self.starting_stream.poll_next_unpin(cx) {
+            if let Some(fut) = fut {
+                self.download_stream.scheduled.push_back(fut);
+            } else {
+                self.download_stream.ended = true;
+            }
+        }
+
+        let poll = self.download_stream.poll_next_unpin(cx);
+
+        if poll.is_ready() {
+            self.starting_stream.downloaded();
+        }
+
+        poll
+    }
+}
+
+type DownloadFuture = Pin<Box<dyn Future<Output = eyre::Result<Box<Path>>>>>;
+
+#[pin_project]
+struct DownloadStream {
+    #[pin]
+    pub downloads: FuturesOrdered<DownloadFuture>,
+    scheduled: VecDeque<DownloadFuture>,
+    max_concurrent_downloads: usize,
+    ended: bool,
+}
+
+impl Debug for DownloadStream {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "DownloadStream({})", self.downloads.len())
+    }
+}
+
+impl Stream for DownloadStream {
+    type Item = eyre::Result<Box<Path>>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        for _ in 0..self.max_concurrent_downloads - self.downloads.len() {
```

Comment:
I see, yeah this should hold  because we can't push more downloads than max_concurrent_downloads